### PR TITLE
GH-2359: Initial AOT Support for Native

### PR DESCRIPTION
**Pull request by garyrussell**
Resolves https://github.com/spring-projects/spring-kafka/issues/2359

This is sufficient for the basic Kafka smoke test.

I will issue another PR if changes are needed for the Streams and Avro versions.
I have a `KafkaAvroBeanRegistrationAotProcessor` but I have omitted it from
this commit because it hasn't been tested.

### CODE REVIEW

File: spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java:225]: I guess it is preferable to use a dedicated for this purposes API on the hints: `registerTypeIfPresent()`

File: spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java:125]: I believe annotations must be registers with their special API like you did in Spring AMQP.

File: spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java:126]: I don't think we need this annotation to expose...

File: spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java:157]: Looks like you have duplicated this type.
I also wonder if we really use it via reflection at all...
In SI I registered it only for serialization.

File: spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java:171]: I wonder if you tried to go baby steps with all of these types registration.
Looks like you just ported whatever there was in Spring Native...

**Review by artembilan — CHANGES_REQUESTED** [review]
I wonder if Avro stuff has to be somewhere else: doesn't look like it is just Kafka-specific...

**Comment by garyrussell** [comment]
The avro processor is specific to spring-kafka because it discovers listeners that use Avro types and registers them for reflection.

I left it out of this PR to avoid any confusion (and because I haven't tested it yet).

File: spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java
Comment (garyrussell) [reply spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java:225]: The problem with that is I have to use FQ class names instead of the classes themselves.

File: spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java
Comment (garyrussell) [reply spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java:171]: That is exactly correct; I would rather get it working quickly and then we can prune it later.

File: spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java
Comment (garyrussell) [reply spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java:157]: Again, I just copied what was in spring-native, for now.

File: spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java:124]: This annotation is registered by Spring Framework by itself. See `MessagingAnnotationsRuntimeHints`

File: spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java:121]: I still don't think we need this. The `@Import` processing is initiated by AOT. So, that our `KafkaBootstrapConfiguration` is already registered at runtime.
Therefore I don't see a reason in this registration.

**Review by artembilan — CHANGES_REQUESTED** [review]

File: spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java:205]: Why FQCN? Look like this class is a part of Java by itself

File: spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java
Comment (artembilan) [reply spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java:211]: Are you going to fix this via `onReachableType(TypeReference.of("org.apache.kafka.streams.StreamsBuilder")` ?

File: spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java:209]: `AopProxyUtils.completeJdkProxyInterfaces()` ?

**Review by artembilan — CHANGES_REQUESTED** [review]

File: spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java
Comment (garyrussell) [reply spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java:205]: Again; copied from spring-native.

File: spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java
Comment (garyrussell) [reply spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java:211]: >That is correct: there is a variant for fqcn and classLoader

Right converting all those classes to FQCN is a pain.

Where is `onReachableType`?

Comment (artembilan) [reply spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaRuntimeHintsRegistrar.java:211]: See how I used it in my SI PR: https://github.com/spring-projects/spring-integration/pull/3832/files#diff-b471878878aa8838786d4a93b19f8014d497db6ace582572f834da814720e01bR97

**Review by artembilan — CHANGES_REQUESTED** [review]

File: spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaAvroBeanRegistrationAotProcessor.java
Comment (garyrussell) [reply spring-kafka/src/main/java/org/springframework/kafka/aot/KafkaAvroBeanRegistrationAotProcessor.java]: Actually, it turns out we don't need to process the annotations, only `GenericMessageListener`s that use Avro types - the ones that are method parameters are already discovered.

**Review by artembilan — APPROVED** [review]
LGTM

Let me know if anything else you'd like to add to this PR or it is good to go for merging!

**Comment by garyrussell** [comment]
Thanks; all smoke tests are green, but I will add a couple more before merging this.

**Comment by garyrussell** [comment]
Ok to merge after this final commit; thanks!

**Pull request merged**
